---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# rprimer <img src='man/figures/rprimer.png' align="right" height="139" />

<!-- badges: start -->
[![R-CMD-check](https://github.com/sofpn/rprimer/workflows/R-CMD-check/badge.svg)](https://github.com/sofpn/rprimer/actions)
<!-- badges: end -->

rprimer is an R package that designs degenerate oligos and PCR assays from a multiple DNA sequence alignment of target sequences of interest. The package is specifically designed for sequence variable viruses. 

## Installation

To install rprimer from [Bioconductor](https://bioconductor.org/packages/devel/bioc/html/rprimer.html), start R (version 4.2) and enter: 

  ``` r
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("rprimer")
```

``` {r, message=FALSE, warn=FALSE, echo=FALSE}
devtools::load_all(".")
```

Attach the package by calling: 

```r
library(rprimer)
```

## Overview 

The package contains five main functions: 

* `consensusProfile()`
* `designOligos()` 
* `designAssays()`
* `checkMatch()`
* `plotData()`

## Shiny application

The package can be run through a Shiny application (a graphical user interface). To start the application, type `runRprimerApp()` from within R upon installing and attaching the package.

The application can also be found online, [here](https://sofpn.shinyapps.io/rprimer).

## Workflow 

### Import alignment

The first step is to import an alignment with target sequences of interest. This is done by using `readDNAMultipleAlignment()`.

The file "example_alignment.txt" contains an alignment of 50 hepatitis E virus sequences. 

```{r}
infile <- system.file("extdata", "example_alignment.txt", package = "rprimer")

myAlignment <- readDNAMultipleAlignment(infile, format = "fasta")
```

### Step 1: `consensusProfile` 

`consensusProfile()` takes a `DNAMultipleAlignment` as input and returns all the information needed for the subsequent design process.

```{r}
myConsensusProfile <- consensusProfile(myAlignment, ambiguityThreshold = 0.05)
```

Results (row 100-106):

```{r, echo=FALSE}
knitr::kable(myConsensusProfile[100:106, ], digits = 2)
```

The results can be visualized with `plotData()`:

```{r, fig.width=12, fig.height=6}
plotData(myConsensusProfile)
```

### Step 2: `designOligos`

The next step is to design oligos. You can either use the default settings as below, or adjust them as preferred (see the package vignette or `?designOligos` for more information). The default settings allow a maximum degeneracy of four, which means that only the most conserved regions of the genome will be considered as oligo binding sites. 

``` {r}
myOligos <- designOligos(myConsensusProfile)
```

Results (first six rows): 

```{r, echo=FALSE}
knitr::kable(head(myOligos), digits = 2)
```

The results can be visualized as a dashboard, using `plotData()`: 

```{r, fig.width=12, fig.height=8}
plotData(myOligos)
```

### Step 3: `designAssays`

`designAssays()` finds pairs of forward and reverse primers and combine them with probes, if probes are present in the input dataset. You can either use the default settings as below, or adjust the design constraints (see the package vignette or `?designAssays` for more information).

```{r}
myAssays <- designAssays(myOligos)
```

Results (first six rows):

```{r, echo=FALSE}
knitr::kable(head(myAssays), digits = 2)
```

The assays can be visualized using `plotData()`: 

```{r, fig.width=12, fig.height=6}
plotData(myAssays)
```

### Additional step: `checkMatch` 

`checkMatch()` shows the proportion and names of the target sequences in the input alignment that match with the generated oligos or assays. See the package vignette or `?checkMatch` for more information. 

```{r}
## Randomly select six oligos to illustrate an example 
selection <- sample(seq_len(nrow(myOligos)), size = 6)

matchTableOligos <- checkMatch(myOligos[selection, ], target = myAlignment)
```

Results:

```{r, echo=FALSE}
knitr::kable(matchTableOligos, digits = 2)
```

The match table can be visualized using `plotData()`: 

```{r, fig.width=12, fig.height=6}
plotData(matchTableOligos)
```

## More information

Please see the [package vignette](https://bioconductor.org/packages/devel/bioc/vignettes/rprimer/inst/doc/getting-started-with-rprimer.html) for more information on how to use the package. 

## Citation

Persson S., Larsson C., Simonsson M., EllstrÃ¶m P. (2022) rprimer: an R/bioconductor package for design of degenerate oligos for sequence variable viruses. [*BMC Bioinformatics* 23:239](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-022-04781-0)
